File: b75/src/main/java/com/jregistr/dsa/arrays/TwoSums.java
package com.jregistr.dsa.arrays;

import java.util.HashMap;

public class TwoSums {
    public int[] twoSum(int[] nums, int target) {
        var prev = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i ++) {
            int num = nums[i];
            int other = target - num;

            if (prev.containsKey(other)) {
                return new int[]{i, prev.get(other)};
            } else {
                prev.put(num, i);
            }
        }

        return new int[]{};
    }
}

File: b75-rs/src/two_sums.rs
use std::collections::HashMap;

struct TwoSums;

impl TwoSums {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {

        let mut prev = HashMap::new();

        for (index, num) in nums.into_iter().enumerate() {
            let other = target - num;
            if let Some(other_index) = prev.get(&other) {
                return vec![*other_index as i32, index as i32];
            } else {
                prev.insert(num, index);
            }
        }

        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test1() {
        assert_eq!(vec![0, 1], TwoSums::two_sum(vec![2,7,11,15], 9));
    }

    #[test]
    fn test2() {
        assert_eq!(vec![0, 1], TwoSums::two_sum(vec![3, 3], 6));
    }

    #[test]
    fn test3() {
        assert_eq!(vec![1, 2], TwoSums::two_sum(vec![3,2,4], 6));
    }
}